etcdserver/api/v3rpc/watch.go

```diff
@@ -336,8 +337,35 @@ func (sws *serverWatchStream) sendLoop() {
 			}
 
 			mvcc.ReportEventReceived(len(evs))
-			if err := sws.gRPCStream.Send(wr); err != nil {
-				return
+			// The grpc package's defaultServerMaxSendMessageSize, defaultServerMaxReceiveMessageSize
+			// and the maxSendMesgSize are set to lower numbers in order to demonstrate
+			// the fragmenting.
+			maxEventsPerMsg := 20
+			watchRespSent := false
+			for !watchRespSent {
+				// There isn't a reasonable way of deciphering the cause of the send error.
+				// One solution that comes to mind is comparing the string error message
+				// to the expected string for a grpc message indicating that the
+				// message size is too large, but this approach would fail if grpc
+				// were to change their error message semantics in the future. The
+				// approach here is to assume that the error is caused by the message
+				// being too large and incrementally increasing the number of fragments
+				// until the maxEventsPerMesg becomes 0 and we are therefore sure
+				// that the error in the send operation was not caused by the message
+				// size.
+				if maxEventsPerMsg == 0 {
+					return
+				}
+				for _, fragment := range FragmentWatchResponse(maxEventsPerMsg, wr) {
```

Comment:
Ok. I'll have to think about how to implement this. Do you think a map between the watchId and fragment boolean that would be stored in the `serverWatchStream` would be a good idea?  It seems like once the  `WatchCreateRequest` reaches the `serverWatchStream`'s `recvLoop` the fragment boolean has to be stored somewhere rather than passed along to the watch request on the mvcc backend.